corehq/apps/cloudcare/views.py

```diff
@@ -212,6 +214,19 @@ def _default_lang():
             "has_geocoder_privs": has_geocoder_privs(domain),
             "valid_multimedia_extensions_map": VALID_ATTACHMENT_FILE_EXTENSION_MAP,
         }
+
+        langs_by_code = {}
+
+        def get_lang_name(lang_code):
+            if not langs_by_code.get(lang_code):
+                langs_by_code[lang_code] = get_name(lang_code)
+            return langs_by_code[lang_code]
+
+        for app in apps:
+            lang_codes = app.get("langs", [])
+            lang_code_name_mapping = {code: get_lang_name(code) for code in lang_codes}
+            app["lang_code_name_mapping"] = lang_code_name_mapping
```

Comment:
I see this is stored on the app object, but what if you made it a top-level variable in `context`?  Then you wouldn't need to lookup the app or pass around the mapping in order to access it - you could just grab it from initial page data where needed in `showFormMenu`

Reply:
Currently I am passing only the language code and its relevant mapping for each app separately. Do you mean to use the global language code mapping at the top level in `context` and use common mapping to extract app specific language names in the `showFormMenu` function? @esoergel

Reply:
For example, imagine a project like this:
* App 1: en, es, fra
* App 2: en, fra, swa

As written, this stores something like:
```python
app1['lang_code_name_mapping'] = {
    'en': 'English',
    'es': 'Spanish',
    'fra': 'French',
}
```
(and likewise for app 2).  What I'm saying is you could instead store one global dict of all languages used in the applications on the project, like this:

```python
{
    'en': 'English',
    'es': 'Spanish',
    'fra': 'French',
    'swa': 'Swahili',
}
```
Then you could extract the app-specific language names in `showFormMenu` as you described.  It'd let you skip the lookup in `controller.js`.

Reply:
Got it. Thanks Ethan for the explanation.

Reply:
https://github.com/dimagi/commcare-hq/pull/33158/commits/7268440ad42d92af29dffd2ba1c43f06acbe19be